### PR TITLE
Issue #5019 - hot-reload SSL certificates if keystore file changed

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #5019**

Added an `ssl-reload` module which uses `Scanner` to monitor the keystore file registered with the `SslContextFactory` and reloads the keystore file if it detects any changes.

### CODE REVIEW

**Comment by sbordet** [comment]
@lachlan-roberts please redo this PR as follows:

* remove the Maven module `jetty-ssl-reload` as it is only 1 class
* move the Jetty module `jetty-ssl-reload.mod` to Maven module `jetty-server`, where the other SSL Jetty modules are
* class and test cases should be in `jetty-util`.
* add a test where the keyStore file is a symlink and change the target of the symlink

**Review by sbordet — CHANGES_REQUESTED** [review]
See PR comment.

**Comment by lachlan-roberts** [comment]
> - remove the Maven module jetty-ssl-reload as it is only 1 class
> - move the Jetty module jetty-ssl-reload.mod to Maven module jetty-server, where the other SSL Jetty modules are
> - class and test cases should be in jetty-util.
> - add a test where the keyStore file is a symlink and change the target of the symlink


@sbordet I did all these, but moved the tests to `test-integration` as it needed a `Server` dependency. Maybe there is a better location for this.

File: jetty-server/src/main/config/etc/jetty-ssl-reload.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
+
+<Configure id="Server" class="org.eclipse.jetty.server.Server">
+
```
Comment (sbordet) [inline jetty-server/src/main/config/etc/jetty-ssl-reload.xml]: Remove empty line!

File: jetty-server/src/main/config/modules/ssl-reload.mod [hunk 2]
```diff
@@ -0,0 +1,21 @@
+DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
```
Comment (joakime) [inline jetty-server/src/main/config/modules/ssl-reload.mod]: This is a comment.
Add `#` to the start please.

File: jetty-server/src/main/config/modules/ssl-reload.mod [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+[tags]
+connector
+ssl
+
+[depend]
+ssl
+
+[lib]
+lib/jetty-ssl-reload-${jetty.version}.jar
```
Comment (sbordet) [inline jetty-server/src/main/config/modules/ssl-reload.mod]: I don't think there is this jar anymore?

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            reload();
+    }
+
+    @Override
+    public void fileRemoved(String filename) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("removed {}", filename);
+
+        // TODO: do we want to do this?
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Feel free to remove this, you have a testcase for it.
This works as-is, and makes sense.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.List;
+
+import org.eclipse.jetty.util.Scanner;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class SslKeyStoreScanner extends AbstractLifeCycle implements Scanner.DiscreteListener
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Perhaps call it `KeyStoreScanner` ?

Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Drop `Ssl` from the class name and extend `ContainerLifeCycle` so we can add the `Scanner` as a bean.

**Review by joakime — CHANGES_REQUESTED** [review]
Some small changes.

File: jetty-server/src/main/config/modules/ssl-reload.mod [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+[xml]
+etc/jetty-ssl-reload.xml
```
Comment (sbordet) [inline jetty-server/src/main/config/modules/ssl-reload.mod]: Rename the XML file to `jetty-ssl-context-reload.xml`.

File: jetty-server/src/main/config/modules/ssl-reload.mod [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+[ini-template]
+# Monitored directory scan period (seconds)
+# jetty.ssl.reload.scanInterval=1
```
Comment (sbordet) [inline jetty-server/src/main/config/modules/ssl-reload.mod]: Properties should be `jetty.sslContext.reload.scanInterval`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    private final SslContextFactory sslContextFactory;
+    private final File keystoreFile;
+    private final List<File> files = new ArrayList<>();
+    private Scanner _scanner;
+    private int _scanInterval = 1;
+
+    public SslKeyStoreScanner(SslContextFactory sslContextFactory)
+    {
+        this.sslContextFactory = sslContextFactory;
+        this.keystoreFile = new File(URI.create(sslContextFactory.getKeyStorePath())); // getKeyStorePath is giving url instead of path?
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Perhaps try ... `sslContextFactory.getKeyStoreResource().getFile()`?

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            LOG.debug("created {}", this);
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug(this.getClass().getSimpleName() + ".doStart()");
+
+        _scanner = new Scanner();
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Make `_scanner` `final` and add it as a bean in the constructor.
Here just configure it and call `super` to start it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java (hunk 4 above)
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: Remove comment as you have a test case for this behavior and seems to work fine.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java (hunk 8 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: How do you handle it when there is no KeyStore Resource/Path/File?

Such as when the user uses `SslContextFactory.setKeyStore(KeyStore)` or `SslContextFactory.setKeyStoreResource(Resource)` on a resource that is not backed by a File?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/config/modules/ssl-reload.mod (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-server/src/main/config/modules/ssl-reload.mod]: None of the other module files have this as a comment? I don't think this is supposed to have a `#`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java (hunk 8 above)
Comment (lachlan-roberts) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java]: If there is no keystore file we will hit one of these `IllegalArgumentException`s and have an error during startup. I think this is fine as its bad configuration to have this when the keystore is not backed by a file.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.Collections;
+
+import org.eclipse.jetty.util.Scanner;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class KeyStoreScanner extends ContainerLifeCycle implements Scanner.DiscreteListener
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java:39]: Add javadocs.

File: jetty-server/src/main/config/etc/jetty-ssl-context-reload.xml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
+
+<Configure id="Server" class="org.eclipse.jetty.server.Server">
+  <Call name="addBean">
+    <Arg>
+      <New id="sslKeyStoreScanner" class="org.eclipse.jetty.util.ssl.KeyStoreScanner">
```
Comment (sbordet) [inline jetty-server/src/main/config/etc/jetty-ssl-context-reload.xml]: Drop `ssl` in the `id`.

**Review by sbordet — CHANGES_REQUESTED** [review]
Please add some documentation in `configuring-ssl.adoc`.

File: jetty-server/src/main/config/modules/ssl-reload.mod (hunk 2 above)
Comment (joakime) [reply jetty-server/src/main/config/modules/ssl-reload.mod]: The other modules are in the wrong.

Put this on your other PR for cleanup.

```
$ git grep "# DO NOT EDIT"
jetty-cdi/src/main/config/modules/cdi-decorate.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-cdi/src/main/config/modules/cdi-spi.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-cdi/src/main/config/modules/cdi.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-cdi/src/main/config/modules/cdi2.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-deploy/src/main/config/modules/decorate.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-home/src/main/resources/modules/setuid.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
jetty-server/src/main/config/modules/https.mod:# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
```

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java [hunk 12]
```diff
@@ -1131,6 +1131,9 @@ public SSLContext getSslContext()
 
         synchronized (this)
         {
+            if (_factory == null)
+                throw new IllegalStateException("reload failed SslContext unavailable");
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Change the message to "SslContextFactory not started or reload failed" in the 3 places you have added it.

We will know from the stack trace what is unavailable.

**Review by sbordet — APPROVED** [review]
Please add *.adoc documentation as well.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java (hunk 12 above)
Comment (lachlan-roberts) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: There is this check just above the synchronized block
```java
if (!isStarted())
    return _setContext;
```
which is why I didn't put the "not started" bit in the exception message, or do you expect this state to be able to change in between this check and the synchronized block?

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Cool, just say "SslContextFactory reload failed" then.

**Review by sbordet — APPROVED** [review]

**Comment by mantryashutosh** [comment]
Does the hot reload work if the keystore file is a soft link ?

**Comment by joakime** [comment]
I wouldn't expect it to, as the soft link itself wasn't updated, meaning there's no file change (last modified) detectable.

**Comment by lachlan-roberts** [comment]
@mantryashutosh it should still work if the keystore file is a soft link, we even have a test for this at `KeyStoreScannerTest.testReloadChangingTargetOfSymbolicLink()`.

**Comment by mantryashutosh** [comment]
that's great, thank you so much @lachlan-roberts . Will test it out

**Pull request merged**